Address PlatformTopo issues around the lscpu cache

bgeltz opened the pull request:
- Resolves #1548.
- Resolves #2269.
- Related to #2256.
- Related to #2257.

bgeltz committed:
Make PlatformTopo secure

- Introduce methods to create cache for the service.
- Create the cache with 644 permissions.
- Rename open_lscpu() to read_lscpu().
- Remove close_lscpu().
- Rework PlatformTopoImp::lscpu to use only C++ string processing.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Implement uptime checking for lscpu cache

- Use move semantics when creating the cache file.
- Remove geopm_topo_popen() from the public API.
- Have the service create the service-specific cache.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -8,17 +8,25 @@
+#include "geopm/Helper.hpp"
 #include "AcceleratorTopo.hpp"
 
 #include "config.h"
```
Unrelated to this patch, but config.h should be first.

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +461,80 @@ namespace geopm
-                              "lscpu -x > $out && "
-                              "chmod a+rw $out";
+        // If cache file is not present, or is too old, create it
+        bool file_ok = false;
```
file_ok -> is_file_okay?

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +461,80 @@ namespace geopm
+                throw Exception("PlatformTopo::create_cache(): Could not create temp file: ",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            close(tmp_fd);
+
+            int err = chmod(tmp_path, perms);
```
```suggestion

            int err = fchmod(tmp_fd, perms);
            close(tmp_fd);
```

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +461,80 @@ namespace geopm
+            }
+
+            std::string cmd = "out=" + std::string(tmp_path) + ";"
+                              "lscpu -x > $out";
```
```suggestion
            std::ostringstream cmd;
            cmd << "lscpu -x >> " << tmp_path << ";"
```

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +461,80 @@ namespace geopm
+
             FILE *pid;
-            int err = geopm_sched_popen(cmd.c_str(), &pid);
+            err = geopm_topo_popen(cmd.c_str(), &pid);
```
```suggestion
            err = geopm_topo_popen(cmd.str().c_str(), &pid);
```

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +461,80 @@ namespace geopm
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+
+            err = rename(tmp_path, cache_file_name.c_str()); // Ensure both files are on the same filesystem
```
think we can drop the comment now.

cmcantalupo commented:
```
In file included from src/PlatformTopo.cpp:8:
./src/PlatformTopoImp.hpp:50:18: error: private field 'm_do_fclose' is not used [-Werror,-Wunused-private-field]
            bool m_do_fclose;
```

cmcantalupo commented on service/src/geopm/PlatformTopo.hpp:
```diff
@@ -17,7 +17,8 @@
     /// @brief Identify host CPU.
-    int geopm_read_cpuid(void); // Only used in examples/geopm_platform_supported.cpp
+    int geopm_read_cpuid(void);
+    int geopm_topo_popen(const char *cmd, FILE **fid);
```
Maybe it would be better to just re-declare this function in the test file that needs it.  Otherwise we are committing to the API for 2.0 since this is an installed header.  I hope we can get away from lscpu eventually.

cmcantalupo reviewed (commented):
Not sure I like having the geopm_topo_popen() in the public API.  Otherwise looks good.

bgeltz committed:
Address review feedback

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

avilcheslopez commented on service/geopmdpy/topo.py:
```diff
@@ -200,14 +200,18 @@ def create_cache():
-    performed.  To force the creation of a new cache file call
-    os.unlink('/tmp/geopm-topo-cache') prior to calling this function.
+    "-rw-r--r--", i.e. 644. The path for the cache file is
+    /run/geopm-service/geopm-topo-cache.  If thie file is older than
```
Minor typo: thie

avilcheslopez commented on service/src/PlatformTopo.cpp:
```diff
@@ -55,9 +63,44 @@ int geopm_read_cpuid(void)
+    err = sigaction(SIGCHLD, &g_popen_complete_signal_action, &save_action);
+    if (!err) {
+        *fid = popen(cmd, "r");
+        while (*fid && !g_is_popen_complete) {
+
+        }
```
It makes me a bit nervous to have this wait until the expected signal is received, especially because the interface this provides allows passing any string as the command for popen(). At the very least, input validation should be performed on cmd, but would also recommend establishing a limit on how long this loops.

This is more of a concern if this function is exposed more publicly. Otherwise, if it's only used internally and input to it is very well controlled, it's much less of a concern.

bgeltz replied:
I have moved the declaration of this out of the public header, so for the moment this is only used in the PlatformTopoImp, and the MSRIOGroupTest.  In both cases the input is very well controlled.

Moreover, this chunk of code was just moved around in this PR.  It is certainly a candidate for a refactor if we wanted to put this back into a public API.

avilcheslopez commented on service/src/PlatformTopo.cpp:
```diff
@@ -55,9 +63,44 @@ int geopm_read_cpuid(void)
+        g_is_popen_complete = 0;
+        sigaction(SIGCHLD, &save_action, NULL);
+    }
+    if (!err && *fid == NULL) {
+        err = errno ? errno : GEOPM_ERROR_RUNTIME;
+    }
```
Minor suggestion: you can move this `if` into the previous `if` and reduce it to:

`if (*fid == NULL) {...`

since the previous `if` already ensures `!err` is true.

avilcheslopez commented on service/src/PlatformTopo.cpp:
```diff
@@ -418,30 +460,79 @@ namespace geopm
+                throw Exception("PlatformTopo::create_cache(): Could not create temp file: ",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            int err = fchmod(tmp_fd, perms);
+            if (err) {
```
In case of error, before throwing: `close(tmp_fd)`

However, seeing the intention is to close it, anyway, you could simply close it right after the call to `fchmod()` regardless of return value.

avilcheslopez commented on service/src/PlatformTopo.cpp:
```diff
@@ -544,77 +635,90 @@ namespace geopm
-                }
-                m_do_fclose = false;
+            mode_t expected_perms;
+            if (file_path.c_str() == M_SERVICE_CACHE_FILE_NAME) {
```
This banks on the fact that when the service cache file name is passed in `file_path`, it will be passed as `M_SERVICE_CACHE_FILE_NAME` (which is how it should be, hence this const member's existence). Maybe I'm being too paranoid, but for full robustness, I would recommend using std::string's operator==.

If not for that, at least for consistency, seeing an earlier similar check is using std::string's operator==, as opposed to doing pointer comparison.

avilcheslopez replied:
Sorry, I'm way off. M_SERVICE_CACHE_FILE_NAME is of type std::string. Even if it had been const char* my comment is incorrect.

The operator== that takes char* and std::string is being used. I guess the only thing worth noting is that calling c_str() is not necessary and both std::strings could be passed to operator==.

avilcheslopez reviewed (commented):
LGTM! Just a few minor comments...

bgeltz committed:
Check the new file contents in the unit test

- Use spoof_lscpu() to avoid trying to parse the real
  lscpu output in CI.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Only create the /tmp cache if the service cache is not available

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Remove create_cache_service()

- This is all handled with create_cache() now.
- Necessary to handle use cases where create_cache() was
  previously called (e.g. launcher.py and geopmread/write).

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I26d378196beb4ca0aee6264d892407ed19ee9b5d
